portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -1212,6 +1357,23 @@ private Key _getKey() {
 		return null;
 	}
 
+	private LiferayMutablePortletParameters _getMutablePortletParameters(
+		String parameterName) {
+
+		if (_lifecycle.equals(PortletRequest.ACTION_PHASE) &&
+			!_mutableRenderParameters.isPublic(parameterName) &&
+			!parameterName.startsWith(
+				PortletQName.PRIVATE_RENDER_PARAMETER_NAMESPACE)) {
+
+			return (LiferayMutablePortletParameters)_mutableActionParameters;
+		}
+		else if (_lifecycle.equals(PortletRequest.RESOURCE_PHASE)) {
+			return (LiferayMutablePortletParameters)_mutableResourceParameters;
+		}
+
+		return _mutableRenderParameters;
+	}
+
 	private void _initMutableRenderParameters() {
```

Comment:
@shuyangzhou The `_initMutableRenderParameters()` method was introduced last in the process. It is a result of trying to correct failures in tests reported by CI. The method is called first in the constructor in order to initialize the mutable render parameters. It is also called by the `setCopyCurrentRenderParameters(boolean)` method because the developer might do something like this:
```
<portlet:renderURL copyCurrentRenderParameters="true>
</portlet:renderURL>
```

When the developer does this, the call to `PortletURL.setCopyCurrentRenderParameters(boolean)` obviously happens *after* the `PortletURLImpl` constructor. Because of this, the mutable render parameters have to be re-initialized accordingly.